### PR TITLE
add error if drop_index is negative

### LINKED ISSUE #7924: handling of negative drop_axis in map_blocks and map_overlap
**What happened**:

If a negative value for `drop_axis` is passed into either `map_blocks` or `map_overlap` a non-informative exception is raised.

**What you expected to happen**:

I would expect this would work as in NumPy for negative axis arguments where axis becomes `axis = axis % array.ndim`. If it is not intended to work, then it should raise a user-friendly `AxisError`. This came up while working on #7894 and it was decided it should be addressed in a separate PR.

I am interested in submitting a PR to fix this, but have a question on intended behavior:
- The `map_blocks` and `map_overlap` functions can take multiple array arguments. If that is the case, what should happen in the case where both arrays do not have the same number of dimensions? (This is important because we need to use ndim in order to convert a negative `drop_axis` into a non-negative one). Should it error in this case, use `ndim` of the first array, or something else? 

**Minimal Complete Verifiable Example**:
```Python
import dask.array as da

drop_axis = (-1,)
x = da.random.standard_normal((5, 10, 8), chunks=(2, 5, 4))
y = da.map_blocks(lambda x: x.mean(axis=drop_axis), x, drop_axis=drop_axis, dtype=float).compute()
```

<details>
<summary>Traceback</summary>

```
gives an `IndexError` with the following extended traceback

Traceback (most recent call last):

  File "/tmp/ipykernel_1277728/1452951085.py", line 9, in <module>
    y = da.map_blocks(_mean, x, drop_axis=drop_axis, dtype=float).compute()

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/base.py", line 285, in compute
    (result,) = compute(self, traverse=False, **kwargs)

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/base.py", line 568, in compute
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/base.py", line 568, in <listcomp>
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/array/core.py", line 1070, in finalize
    return concatenate3(results)

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/array/core.py", line 4799, in concatenate3
    chunks = chunks_from_arrays(arrays)

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/array/core.py", line 4583, in chunks_from_arrays
    result.append(tuple([shape(deepfirst(a))[dim] for a in arrays]))

  File "/home/lee8rx/mambaforge/envs/skimage-dev/lib/python3.8/site-packages/dask/array/core.py", line 4583, in <listcomp>
    result.append(tuple([shape(deepfirst(a))[dim] for a in arrays]))

IndexError: tuple index out of range
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2021.06.2
- Python version: 3.8
- Operating System: linux
- Install method (conda, pip, source): conda-forge

### PR DESCRIPTION
**Pull request by spiyer99**
Resolves #7924.

I added a ValueError if the `drop_index` is less than `0`. 

Doesn't change too much. This is my first contribution to this repository. 

Let me know what you think. Happy to make changes!

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jsignell** [comment]
@dask/gpu can someone trigger tests?

File: dask/array/core.py [hunk 1]
```diff
@@ -694,6 +694,8 @@ def map_blocks(
         dtype = apply_infer_dtype(func, args, original_kwargs, "map_blocks")
 
     if drop_axis:
+        if any(i < 0 for i in drop_axis): 
+            raise ValueError("Drop_axis is {0}, expected axis to be non-negative.".format(drop_axis))
```
Comment (jsignell) [inline dask/array/core.py:698]: Nitpick, but I find f strings easier to read.

```suggestion
            raise ValueError(f"Drop_axis is {drop_axis}, expected axis to be non-negative.")
```

**Review by jsignell — COMMENTED** [review]
Thanks for opening this @spiyer99! I just have a small comment

**Comment by jacobtomlinson** [comment]
ok to test

File: dask/array/core.py (hunk 1 above)
Comment (spiyer99) [reply dask/array/core.py:698]: Thanks @jsignell - I'll add these changes in. Sorry for the late reply btw!

**Comment by spiyer99** [comment]
Thanks for the help @jsignell! I added your changes in.

**Comment by jsignell** [comment]
It looks like you just have a linting issue. You can refer to https://docs.dask.org/en/latest/develop.html#code-formatting for info on how to run the code formatting checks.

**Comment by spiyer99** [comment]
@jsignell, thanks for your help. Sorry, I’m a bit confused. The code formatting checks are passing locally. The CI code reformatter seems to have deleted a line, only to add it back in:

Screen Shot 2021-09-05 at 7 04 32 pm

It seems like a whitespace issue or something? What do you think?

**Comment by jsignell** [comment]
It's hard to see on that diff, but there was a space at the end of the line that the linter is complaining about, that is what is being removed with black.

I can see it on the diff in vscode when I pull this branch and run `pre-commit run --all-files`. 

image


Just let me know if you want me to push to your branch to get the linting issue fixed up.

**Comment by spiyer99** [comment]
> It's hard to see on that diff, but there was a space at the end of the line that the linter is complaining about, that is what is being removed with black.
> 
> I can see it on the diff in vscode when I pull this branch and run `pre-commit run --all-files`.
> 
> image
> 
> Just let me know if you want me to push to your branch to get the linting issue fixed up.

@jsignell - happy for you to push my branch to get the linting issue fixed up. It would be cool to get this done!

**Comment by jsignell** [comment]
Ok I just pushed.

**Comment by jsignell** [comment]
Thanks @spiyer99 for opening this pull request! Congrats on your first Dask PR :)

**Pull request merged**